crates/data_components/src/lib.rs

```diff
@@ -66,3 +71,25 @@ pub trait Stream: Send + Sync {
         table_reference: OwnedTableReference,
     ) -> Result<Arc<dyn TableProvider + 'static>, Box<dyn Error + Send + Sync>>;
 }
+
+#[async_trait]
+pub trait DeleteTableProvider: TableProvider {
+    async fn delete_from(
+        &self,
+        _state: &SessionState,
+        _filters: &[Expr],
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        Err(DataFusionError::Plan("Not implemented".to_string()))
+    }
+}
+
+// There is no good way to allow inter trait casting yet as TableProvider is not controlled
+pub fn cast_to_deleteable<'a>(
+    from: &'a dyn TableProvider,
+) -> Option<&'a (dyn DeleteTableProvider + 'a)> {
+    if let Some(p) = from.as_any().downcast_ref::<MemTable>() {
```

Comment:
Shouldn't this downcast to a `dyn DeleteTableProvider`, not a `MemTable`?

Reply:
essentially this will try to downcast to every concrete type of delete table provider, and return when succeed.

Reply:
Right, but what I meant was we could have an `Adaptor` struct which just wraps the `dyn` trait similar to https://github.com/datafusion-contrib/datafusion-federation/blob/main/datafusion-federation/src/analyzer.rs#L152

Reply:
why it can does that because when the instance is initialized (even with the wrapper), the table source type is already using the trait of source in the table provider. But what we are doing here is convert a tableprovider struct to a new trait type that has a new method.

Reply:
Above example is not converting one type trait to another type of trait. It is consistent through the caller chain.

What we are doing is convert a trait to another trait; which rust won't allow in downcasting.

Reply:
Right, what they have is:

```rust
pub struct FederatedTableProviderAdaptor {
    pub source: Arc<dyn FederatedTableSource>,
}
```

Downcast to the concrete `FederatedTableProviderAdaptor` and return the embedded `source`. We could do something similar.

Reply:
even that, the `source` still need to be using the same type of trait. Can't trick rust in this case.

Reply:
This seems to work: https://github.com/spiceai/spiceai/pull/1145